can/scripts/viewer.py

```diff
@@ -0,0 +1,602 @@
+#!/usr/bin/python
+# coding: utf-8
+#
+# Copyright (C) 2018 Kristian Sloth Lauszus. All rights reserved.
+#
+# Contact information
+# -------------------
+# Kristian Sloth Lauszus
+# Web      :  http://www.lauszus.com
+# e-mail   :  lauszus@gmail.com
+
+from __future__ import absolute_import, print_function
+
+import argparse
+import can
+import curses
+import os
+import six
+import struct
+import sys
+
+from curses.ascii import ESC as KEY_ESC, SP as KEY_SPACE
+from typing import Dict, List, Tuple, Union
+
+from can import __version__
+
+# CANopen function codes
+CANOPEN_NMT = 0x000
+CANOPEN_SYNC_EMCY = 0x080
+CANOPEN_TIME = 0x100
+CANOPEN_TPDO1 = 0x180
+CANOPEN_RPDO1 = 0x200
+CANOPEN_TPDO2 = 0x280
+CANOPEN_RPDO2 = 0x300
+CANOPEN_TPDO3 = 0x380
+CANOPEN_RPDO3 = 0x400
+CANOPEN_TPDO4 = 0x480
+CANOPEN_RPDO4 = 0x500
+CANOPEN_SDO_TX = 0x580
+CANOPEN_SDO_RX = 0x600
+CANOPEN_HEARTBEAT = 0x700
+CANOPEN_LSS_TX = 0x7E4
+CANOPEN_LSS_RX = 0x7E5
+
+# Mask for extracting the CANopen function code
+CANOPEN_FUNCTION_CODE_MASK = 0x780
+
+# Mask for extracting the CANopen node ID
+CANOPEN_NODE_ID_MASK = 0x07F
+
+# CANopen function codes, all the messages except the TPDOx and RPDOx message have a fixed length according to the
+# specs,  so this is checked as well in order to varify that it is indeed a CANopen message
+canopen_function_codes = {
+    CANOPEN_NMT:        {2: 'NMT'},        # Network management (NMT) node control. The node id should not be added to this value
+    CANOPEN_SYNC_EMCY:  {0: 'SYNC',        # Synchronization (SYNC) protocol. The node id should not be added to this value
+                         8: 'EMCY'},       # Emergency (EMCY) protocol
+    CANOPEN_TIME:       {6: 'TIME'},       # Time (TIME) protocol. The node id should not be added to this value
+    CANOPEN_TPDO1:          'TPDO1',       # 1. Transmit Process Data Object (PDO)
+    CANOPEN_RPDO1:          'RPDO1',       # 1. Receive Process Data Object (PDO)
+    CANOPEN_TPDO2:          'TPDO2',       # 2. Transmit Process Data Object (PDO)
+    CANOPEN_RPDO2:          'RPDO2',       # 2. Receive Process Data Object (PDO)
+    CANOPEN_TPDO3:          'TPDO3',       # 3. Transmit Process Data Object (PDO)
+    CANOPEN_RPDO3:          'RPDO3',       # 3. Receive Process Data Object (PDO)
+    CANOPEN_TPDO4:          'TPDO4',       # 4. Transmit Process Data Object (PDO)
+    CANOPEN_RPDO4:          'RPDO4',       # 4. Receive Process Data Object (PDO)
+    CANOPEN_SDO_TX:     {8: 'SDO_TX'},     # Synchronization Object (SYNC) transmit
+    CANOPEN_SDO_RX:     {8: 'SDO_RX'},     # Synchronization Object (SYNC) receive
+    CANOPEN_HEARTBEAT:  {1: 'HEARTBEAT'},  # Network management (NMT) node monitoring
+    CANOPEN_LSS_TX:     {8: 'LSS_TX'},     # Layer Setting Services (LSS) transmit
+    CANOPEN_LSS_RX:     {8: 'LSS_RX'},     # Layer Setting Services (LSS) receive
+}
+
+
+class CanViewer:
+
+    def __init__(self, stdscr, bus, data_structs, ignore_canopen, testing=False):
```

Comment:
You could also add `curses.use_default_colors()` somewhere to allow the terminal window to choose a fitting background instead of always black.